relay-quotas/src/redis.rs

```diff
@@ -346,12 +346,25 @@ impl<T: GlobalLimiter> RedisRateLimiter<T> {
         // client across await points, otherwise it might be held for too long, and we will run out
         // of connections.
         let mut connection = self.client.get_connection().await?;
-        let rejections: Vec<bool> = invocation
+        let results: Vec<i64> = invocation
             .invoke_async(&mut connection)
             .await
             .map_err(RedisError::Redis)?;
 
-        for (quota, is_rejected) in tracked_quotas.iter().zip(rejections) {
+        // TODO: maybe we should only accept i64, but also needs to be positive.
+        let quantity = i64::try_from(quantity).unwrap_or(i64::MAX);
+        for (quota, consumed) in tracked_quotas.iter().zip(results) {
+            let limit = quota.limit();
+
+            // Limit must not be unlimited (-1) and there must not be enough remaining in the quota.
+            //
+            // Note: this logic is synchronized to the one in the Redis script.
+            let is_rejected = limit >= 0
+                && match quantity == 0 || over_accept_once {
+                    true => consumed >= limit,
+                    false => consumed + quantity > limit,
+                };
```

Comment:
I find this expression very difficult to understand. I believe it's equivalent to
```suggestion
            let is_rejected = limit >= 0
                && match over_accept_once {
                    true => consumed >= limit,
                    false => consumed + quantity > limit,
                };
```
(if `quantity` is 0, both branches are equivalent¹; if `quantity` is not 0, `quantity == 0 || over_accept_once` is equivalent to `over_accept_once`).

¹ Apart from the `>=` vs. `>`, which is possibly unintentional?

Reply:
> I believe it's equivalent to

It is not (as you later state).

> ¹ Apart from the >= vs. >, which is possibly unintentional?

This is intentional.

See the comment above: 
```rs
            // Note: this logic is synchronized to the one in the Redis script.
```

And the logic in the lua script:

```lua
        -- Without over_accept_once, we never increment past the limit. if quantity is 0, check instead if we reached limit.
        -- With over_accept_once, we only reject if the previous update already reached the limit.
        -- This way, we ensure that we increment to or past the limit at some point,
        -- such that subsequent checks with quantity=0 are actually rejected.
        --
        -- NOTE: redis-rs crate since version 0.18.0 (2020-12-03) passes '1' in case of true and '0' when false.
        if quantity == 0 or over_accept_once == '1' then
            failed = consumed >= limit
        else
            failed = consumed + quantity > limit
        end
```

There is also a test validating that behaviour.

The method documentation also describes it as:

```rs
    /// A `quantity` of `0` can be used to check if the quota limit has been reached or exceeded
    /// without incrementing it in the success case. This is useful for checking quotas in a different
    /// data category.
```

That's why a 0 quantity needs to fail.

Reply:
Thanks for the explanation. It's clear now that the expression for `is_rejected` models the behavior from the script. I still don't entirely understand it, but I believe there's a reason why it's that way.

Reply:
Discussed offline, tldr; `0` quantity with an exhausted quota should be treated as a rejection, that's why we need `>=`.

Gonna add a comment and look into a way to possibly better express this.